File: rs_manual_teleop/rs_network/src/video_decoder.py
#!/usr/bin/env python3
import cv2

# Pipeline otimizado para latência zero
port = 5006
pipeline = (
    f"udpsrc port={port} caps=\"application/x-rtp, media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264, payload=(int)96\" ! "
    f"rtph264depay ! avdec_h264 ! videoconvert ! appsink sync=false drop=true max-buffers=1"
)

print(f"A iniciar Decoder na porta {port}...")
cap = cv2.VideoCapture(pipeline, cv2.CAP_GSTREAMER)

if not cap.isOpened():
    print("ERRO: Falha ao escutar a porta UDP. Verifica o GStreamer.")
    exit()

print("A aguardar vídeo... (Pressiona 'q' na janela para sair)")

# Loop contínuo sem bloqueios de timers
while True:
    ret, frame = cap.read()
    
    if ret:
        cv2.imshow("Teleop Camera Stream", frame)
        
    # Atualiza a janela e verifica se a tecla 'q' foi pressionada
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

File: rs_manual_teleop/rs_network/src/video_decoder.cpp
#include <rclcpp/rclcpp.hpp>
#include <opencv2/opencv.hpp>
#include <thread>

class VideoDecoderRX : public rclcpp::Node {
public:
    VideoDecoderRX() : Node("video_decoder_rx_cpp") {
        RCLCPP_INFO(this->get_logger(), "A inicializar descodificador C++...");
        
        // Iniciar a leitura de vídeo numa thread separada para latência zero
        video_thread_ = std::thread(&VideoDecoderRX::receive_and_display, this);
    }

    ~VideoDecoderRX() {
        if (video_thread_.joinable()) {
            video_thread_.join();
        }
    }

private:
    void receive_and_display() {
        // Pipeline equivalente à lógica da TUM, mas adaptado para UDP e appsink do OpenCV
    
        std::string pipeline =
            "udpsrc port=5006 buffer-size=2129920 caps=\"application/x-rtp, media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264, payload=(int)96\" ! "
            "rtpjitterbuffer latency=50 ! "
            "rtph264depay ! h264parse ! avdec_h264 output-corrupt=false ! "
            "videoconvert ! appsink sync=false drop=true max-buffers=1";

        cv::VideoCapture cap(pipeline, cv::CAP_GSTREAMER);

        if (!cap.isOpened()) {
            RCLCPP_ERROR(this->get_logger(), "Falha ao abrir a porta UDP 5006 via GStreamer.");
            return;
        }

        RCLCPP_INFO(this->get_logger(), "A escutar vídeo em UDP (Porta 5006)... A janela abrirá quando o veículo transmitir.");

        cv::Mat frame;
        // Loop contínuo independente dos timers do ROS 2
        while (rclcpp::ok()) {
            if (cap.read(frame) && !frame.empty()) {
                cv::imshow("PC B - Teleop Stream (Zero Latency)", frame);
                
                // Atualiza o ecrã a cada 1 milissegundo
                if (cv::waitKey(1) == 27) { // Pressionar 'ESC' para sair
                    break;
                }
            }
        }

        cap.release();
        cv::destroyAllWindows();
    }

    std::thread video_thread_;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<VideoDecoderRX>());
    rclcpp::shutdown();
    return 0;
}